data/content_scripts/visit.js

```diff
@@ -0,0 +1,29 @@
+var SERVER_URL = "https://chronicle.dev.mozaws.net";
+var CREATE_VISIT_ENDPOINT = SERVER_URL + "/v1/visits";
+
+function Visit(url, title, visitedAt) {
+  this.url = url
+  this.title = title;
+  this.visitedAt = visitedAt || new Date().toISOString();
+}
+
+Visit.prototype.save = function() {
+  var deferred = P.defer();
+  var xhr = new XMLHttpRequest();
+  var body = JSON.stringify({ url: this.url, title: this.title, visitedAt: this.visitedAt });
+  xhr.open("POST", CREATE_VISIT_ENDPOINT, true);
+  xhr.setRequestHeader("Content-type", "application/json");
+  xhr.onreadystatechange = function() {
+    if(xhr.readyState === 4) {
+      if(xhr.status === 200) {
+        deferred.resolve(JSON.parse(xhr.responseText));
+      }
+      else if(xhr.status >= 400) {
+        deferred.reject(xhr.responseText);
+      }
```

Comment:
Hmm, interesting question. I'm going to iterate on the API, maybe this upcoming sprint, so it's a timely question, too.

We aren't using 201 Created or 202 Accepted anywhere, so there's no 2xx to worry about other than 200.

This code isn't doing anything fancy with HTTP caching headers, like explicitly sending down an ETag or Last-Modified header, so we don't have to worry about hapi spitting back a 304 Not-Modified. The [xhr spec](https://dvcs.w3.org/hg/xhr/raw-file/tip/Overview.html#infrastructure-for-the-send%28%29-method) says that 302 Redirects should be transparently followed, so I think we're good there; that's really future-proofing anyway, because our API endpoints don't have any redirects. There is a known bug with xhr 302s not being followed on non-same domain requests, but that also doesn't affect us.

All in all, I think this is fine.